src/zodbverify/verify_oid.py

```diff
@@ -13,6 +20,173 @@
+                continue
+            self.inspect_reference_tree(ref, level=level, verbose=verbose)
+
+    @property
+    def root_oid(self):
+        return self.connection.root._root._p_oid
+
+    def oid_or_repr_to_oid(self, oid_or_repr):
+        if isinstance(oid_or_repr, bytes):
+            return oid_or_repr
+        return repr_to_oid(oid_or_repr)
+
+    def oid_or_repr_to_repr(self, oid_or_repr):
+        if isinstance(oid_or_repr, bytes):
+            return oid_repr(oid_or_repr)
+        return oid_or_repr
+
+    # Do not use cache decorators, let ZODB do its caching.
+    def get_obj(self, oid):
+        u"""Get the object from its `oid'."""
+        oid = self.oid_or_repr_to_oid(oid)
+        obj = self.connection.get(oid)
+        obj._p_activate()
+        return obj
+
+    # @instance.memoize
+    def get_obj_as_str(self, oid):
+        try:
+            return str(self.get_obj(oid))
+        except Exception:
+            return '<error>'
+
+    # @instance.memoize
+    def get_physical_path(self, oid):
+        try:
+            return self.get_obj(oid).getPhysicalPath()
+        except Exception:
+            return None
+
+    # @instance.memoize
+    def get_id(self, oid):
+        obj = self.get_obj(oid)
+        if oid == self.root_oid:
+            return 'Root'
+        getId = getattr(obj, 'getId', None)
+        if getId:
+            try:
+                return getId()
+            except:  # noqa
+                pass
+        return getattr(obj, 'id', None)
+
+    # @instance.memoize
+    def get_attr_name(self, oid, parent_oid):
+        oid = self.oid_or_repr_to_oid(oid)
+        obj = self.get_obj(oid)
+        parent = self.get_obj(parent_oid)
+        names_and_values = ((name, getattr(parent, name, None)) for name in dir(parent))
+        return next((name for (name, value) in names_and_values if value is obj), None)
+
+    # @instance.memoize
+    def get_id_or_attr_name(self, oid, parent_oid=None):
+        identifier = self.get_id(oid)
+        if identifier:
+            return identifier
+
+        return self.get_attr_name(oid, parent_oid) if parent_oid else None
+
+    def load_reference_tree(self):
+        path = self._get_reference_cache_path()
+        if os.path.exists(path):
+            with open(path, 'r') as f:
+                logger.info('Loading json reference-cache from {}'.format(path))
+                self.refs = json.load(f)
+        else:
+            self._build_ref_tree()
+            self._store_reference_cache()
+
+    def _store_reference_cache(self):
+        path = self._get_reference_cache_path()
+        dirname = os.path.dirname(path)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+        with open(path, 'w') as f:
+            json.dump(self.refs, f)
+        logger.info('Save reference-cache as {}'.format(path))
+
+    def _get_reference_cache_path(self):
+        cache_dir = os.path.join(os.path.expanduser('~'), '.cache', 'zodbverify')
```

Comment:
So there is no clean-up? This can add up to some data given large databases. IMO keeping old caches is not needed, just wipe them if a unknown transaction id is coming in.